db/background_mgr_attachment_migration.go

```diff
@@ -0,0 +1,351 @@
+		// process from scratch with a new migration ID. Otherwise, we should resume with the migration ID, stats specified in the doc.
+		if statusDoc.State == BackgroundProcessStateCompleted || err != nil {
+			return newRunInit()
+		}
+		a.MigrationID = statusDoc.MigrationID
+		a.SetStatus(statusDoc.DocsChanged, statusDoc.DocsProcessed)
+		a.SetCollectionIDs(statusDoc.CollectionIDs)
+
+		base.InfofCtx(ctx, base.KeyAll, "Attachment Migration: Attempting to resume migration with migration ID: %s", a.MigrationID)
+
+		return nil
+	}
+
+	return newRunInit()
+}
+
+func (a *AttachmentMigrationManager) Run(ctx context.Context, options map[string]interface{}, persistClusterStatusCallback updateStatusCallbackFunc, terminator *base.SafeTerminator) error {
+	db := options["database"].(*Database)
+	migrationLoggingID := "Migration: " + a.MigrationID
+
+	persistClusterStatus := func() {
+		err := persistClusterStatusCallback(ctx)
+		if err != nil {
+			base.WarnfCtx(ctx, "[%s] Failed to persist cluster status on-demand for attachment migration operation: %v", migrationLoggingID, err)
+		}
+	}
+	defer persistClusterStatus()
+
+	var processFailure error
+	failProcess := func(err error, format string, args ...interface{}) bool {
+		processFailure = err
+		terminator.Close()
+		base.WarnfCtx(ctx, format, args...)
+		return false
+	}
+
+	callback := func(event sgbucket.FeedEvent) bool {
+		docID := string(event.Key)
+		collection := db.CollectionByID[event.CollectionID]
+		base.TracefCtx(ctx, base.KeyAll, "[%s] Received DCP event %d for doc %v", migrationLoggingID, event.Opcode, base.UD(docID))
+
+		// Ignore documents without xattrs, to avoid processing unnecessary documents
+		if event.DataType&base.MemcachedDataTypeXattr == 0 {
+			return true
+		}
+
+		// Don't want to process raw binary docs
+		// The binary check should suffice but for additional safety also check for empty bodies
+		if event.DataType == base.MemcachedDataTypeRaw || len(event.Value) == 0 {
+			return true
+		}
+
+		// We only want to process full docs. Not any sync docs.
+		if strings.HasPrefix(docID, base.SyncDocPrefix) {
+			return true
+		}
+
+		a.DocsProcessed.Add(1)
+		syncData, _, _, err := UnmarshalDocumentSyncDataFromFeed(event.Value, event.DataType, collection.userXattrKey(), false)
+		if err != nil {
+			failProcess(err, "error unmarshaling document %s: %v", base.UD(docID), err)
+		}
+
+		if syncData == nil || syncData.Attachments == nil {
+			// no attachments to migrate
+			return true
+		}
+
+		collCtx := collection.AddCollectionContext(ctx)
+		collWithUser := &DatabaseCollectionWithUser{
+			DatabaseCollection: collection,
+		}
+		// xattr migration to take place
+		err = collWithUser.MigrateAttachmentMetadata(collCtx, docID, event.Cas, syncData)
+		if err != nil {
+			failProcess(err, "error migrating document attachment metadata for doc: %s: %v", base.UD(docID), err)
+		} else {
+			a.DocsChanged.Add(1)
+		}
```

Comment:
nit: for readability

```suggestion
		}
		a.DocsChanged.Add(1)
```